pilot/pkg/config/kube/crdclient/cache_handler.go

```diff
@@ -0,0 +1,114 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crdclient
+
+import (
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"  // import GKE cluster authentication plugin
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc" // import OIDC cluster authentication plugin, e.g. for Tectonic
+
+	"istio.io/pkg/log"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config/schema/collection"
+)
+
+// cacheHandler abstracts the logic of an informer with a set of handlers. Handlers can be added at runtime
+// and will be invoked on each informer event.
+type cacheHandler struct {
+	client   *Client
+	informer cache.SharedIndexInformer
+	handlers []func(model.Config, model.Config, model.Event)
+	schema   collection.Schema
+	lister   func(namespace string) cache.GenericNamespaceLister
+}
+
+func (h *cacheHandler) onEvent(old interface{}, curr interface{}, event model.Event) error {
+	if err := h.client.checkReadyForEvents(curr); err != nil {
+		return err
+	}
+
+	var currItem, oldItem runtime.Object
+
+	ok := false
+
+	if currItem, ok = curr.(runtime.Object); !ok {
+		log.Warnf("New Object can not be converted to runtime Object %v, is type %T", curr, curr)
+		return nil
+	}
+	currConfig := TranslateObject(currItem, h.schema.Resource().GroupVersionKind(), h.client.domainSuffix)
+
+	var oldConfig model.Config
+	if old != nil {
+		if oldItem, ok = old.(runtime.Object); !ok {
+			log.Warnf("Old Object can not be converted to runtime Object %v, is type %T", old, old)
+			return nil
+		}
+		oldConfig = *TranslateObject(oldItem, h.schema.Resource().GroupVersionKind(), h.client.domainSuffix)
+	}
+
+	for _, f := range h.handlers {
+		f(oldConfig, *currConfig, event)
+	}
+	return nil
+}
+
+func createCacheHandler(cl *Client, schema collection.Schema, i informers.GenericInformer) *cacheHandler {
+	h := &cacheHandler{
+		client:   cl,
+		schema:   schema,
+		informer: i.Informer(),
+	}
+	h.lister = func(namespace string) cache.GenericNamespaceLister {
+		if schema.Resource().IsClusterScoped() {
+			return i.Lister()
+		}
+		return i.Lister().ByNamespace(namespace)
+	}
+	kind := schema.Resource().Kind()
+	i.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			incrementEvent(kind, "add")
+			cl.tryLedgerPut(obj, kind)
+			cl.queue.Push(func() error {
+				return h.onEvent(nil, obj, model.EventAdd)
+			})
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			cl.tryLedgerPut(cur, kind)
```

Comment:
@therealmitchconnors ,why do we need to update the ledge if the object doesn't change? Should this be moved to line 97?

Reply:
Good point, I would agree, will see what Mitch things. I copied this from the old code btw

Reply:
opened https://github.com/istio/istio/issues/25091 to track this